File: symfony/src/Service/GeneralTestService.php
<?php

namespace App\Service;

use App\Entity\GeneralAnswer;
use App\Entity\GeneralQuestion;
use App\Entity\GeneralTest;
use App\Repository\GeneralAnswerRepository;
use App\Repository\GeneralQuestionRepository;
use App\Repository\GeneralTestRepository;
use Doctrine\ORM\EntityManagerInterface;

class GeneralTestService
{
    public function __construct(
        private EntityManagerInterface $entityManager,
        private GeneralTestRepository $testRepository,
        private GeneralQuestionRepository $questionRepository,
        private GeneralAnswerRepository $answerRepository
    ) {
    }

    public function getAllTests(): array
    {
        return $this->testRepository->findBy([], ['createdAt' => 'DESC']);
    }

    public function getTestById(int $id): ?GeneralTest
    {
        $test = $this->testRepository->find($id);

        if (!$test) {
            throw new \Exception('Test général introuvable');
        }

        return $test;
    }

    public function getQuestionsByTest(GeneralTest $test): array
    {
        return $this->questionRepository->findBy(['test' => $test], ['questionOrder' => 'ASC']);
    }

    public function getAnswersByQuestion(GeneralQuestion $question): array
    {
        return $this->answerRepository->findBy(['question' => $question], ['answerOrder' => 'ASC']);
    }

    public function getQuestionsWithAnswers(GeneralTest $test): array
    {
        $questions = $this->getQuestionsByTest($test);
        $result = [];

        foreach ($questions as $question) {
            $result[] = [
                'question' => $question,
                'answers' => $this->getAnswersByQuestion($question),
            ];
        }

        return $result;
    }

    public function createTest(array $data): GeneralTest
    {
        $title = trim($data['title'] ?? '');
        $description = trim($data['description'] ?? '');
        $createdBy = (int) ($data['created_by'] ?? 1);
        $questionsData = $data['questions'] ?? [];

        if (empty($title)) {
            throw new \Exception('Le titre est obligatoire');
        }

        if (is_numeric($title)) {
            throw new \Exception('Le titre ne doit pas être uniquement des nombres');
        }

        if (strlen($title) < 3) {
            throw new \Exception('Le titre doit contenir au moins 3 caractères');
        }

        if (empty($questionsData)) {
            throw new \Exception('Le test doit contenir au moins une question');
        }

        $test = new GeneralTest();
        $test->setTitle($title);
        $test->setDescription($description ?: null);
        $test->setStatus('DRAFT');
        $test->setCreatedBy($createdBy);

        foreach ($questionsData as $questionData) {
            $questionText = trim($questionData['text'] ?? '');
            $questionOrder = (int) ($questionData['order'] ?? 0);
            $answersData = $questionData['answers'] ?? [];

            if (empty($questionText)) {
                continue;
            }

            $question = new GeneralQuestion();
            $question->setTest($test);
            $question->setQuestionText($questionText);
            $question->setQuestionOrder($questionOrder > 0 ? $questionOrder : 1);

            $answerOrder = 1;
            foreach ($answersData as $answerData) {
                $answerText = trim($answerData['text'] ?? '');
                $score = (int) ($answerData['score'] ?? 0);

                if (empty($answerText)) {
                    continue;
                }

                $answer = new GeneralAnswer();
                $answer->setQuestion($question);
                $answer->setAnswerLabel(chr(64 + $answerOrder));
                $answer->setAnswerText($answerText);
                $answer->setScore($score);
                $answer->setAnswerOrder($answerOrder);

                $question->addAnswer($answer);
                $answerOrder++;
            }

            $test->addQuestion($question);
        }

        if ($test->getQuestions()->isEmpty()) {
            throw new \Exception('Le test doit contenir au moins une question valide');
        }

        $this->entityManager->persist($test);
        $this->entityManager->flush();

        return $test;
    }

    public function updateTest(GeneralTest $test, array $data): GeneralTest
    {
        $title = trim($data['title'] ?? '');
        $description = trim($data['description'] ?? '');
        $status = trim($data['status'] ?? 'DRAFT');

        if (empty($title)) {
            throw new \Exception('Le titre est obligatoire');
        }

        if (is_numeric($title)) {
            throw new \Exception('Le titre ne doit pas être uniquement des nombres');
        }

        if (strlen($title) < 3) {
            throw new \Exception('Le titre doit contenir au moins 3 caractères');
        }

        if (!in_array($status, ['DRAFT', 'ACTIVE', 'INACTIVE', 'ARCHIVED'])) {
            throw new \Exception('Statut invalide');
        }

        $test->setTitle($title);
        $test->setDescription($description ?: null);
        $test->setStatus($status);
        $test->setUpdatedAt(new \DateTime());

        $this->entityManager->flush();

        return $test;
    }

    public function deleteTest(GeneralTest $test): void
    {
        $this->entityManager->remove($test);
        $this->entityManager->flush();
    }
}

File: integrated-project/mindboost-javafx/src/main/java/com/mindboost/app/service/GeneralTestService.java
package com.mindboost.app.service;

import com.mindboost.app.config.DatabaseManager;
import com.mindboost.app.model.GeneralAnswer;
import com.mindboost.app.model.GeneralQuestion;
import com.mindboost.app.model.GeneralTest;
import jakarta.persistence.EntityManager;
import jakarta.persistence.TypedQuery;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class GeneralTestService extends BaseService<GeneralTest> {
    public GeneralTestService(DatabaseManager databaseManager) {
        super(databaseManager, GeneralTest.class);
    }

    public List<GeneralTest> findAllOrdered() {
        try (EntityManager em = databaseManager.createEntityManager()) {
            return em.createQuery("select t from GeneralTest t order by t.createdAt desc", GeneralTest.class)
                .getResultList();
        }
    }

    public GeneralTest requireById(int id) {
        return findById(id).orElseThrow(() -> new IllegalArgumentException("Test général introuvable"));
    }

    public List<GeneralQuestion> getQuestionsByTest(GeneralTest test) {
        try (EntityManager em = databaseManager.createEntityManager()) {
            TypedQuery<GeneralQuestion> query = em.createQuery(
                "select q from GeneralQuestion q where q.test = :test order by q.questionOrder asc",
                GeneralQuestion.class
            );
            query.setParameter("test", test);
            return query.getResultList();
        }
    }

    public List<GeneralAnswer> getAnswersByQuestion(GeneralQuestion question) {
        try (EntityManager em = databaseManager.createEntityManager()) {
            TypedQuery<GeneralAnswer> query = em.createQuery(
                "select a from GeneralAnswer a where a.question = :question order by a.answerOrder asc",
                GeneralAnswer.class
            );
            query.setParameter("question", question);
            return query.getResultList();
        }
    }

    public List<QuestionWithAnswers> getQuestionsWithAnswers(GeneralTest test) {
        List<QuestionWithAnswers> result = new ArrayList<>();
        for (GeneralQuestion question : getQuestionsByTest(test)) {
            result.add(new QuestionWithAnswers(question, getAnswersByQuestion(question)));
        }
        return result;
    }

    public GeneralTest createTest(String title, String description, int createdBy, List<QuestionDraft> questions) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Le titre est obligatoire");
        }
        String trimmedTitle = title.trim();
        if (trimmedTitle.matches("\\d+")) {
            throw new IllegalArgumentException("Le titre ne doit pas être uniquement des nombres");
        }
        if (trimmedTitle.length() < 3 || trimmedTitle.length() > 255) {
            throw new IllegalArgumentException("Le titre doit avoir entre 3 et 255 caractères");
        }
        if (questions == null || questions.isEmpty()) {
            throw new IllegalArgumentException("Le test doit contenir au moins une question");
        }

        GeneralTest test = new GeneralTest();
        test.setTitle(trimmedTitle);
        test.setDescription(description == null || description.isBlank() ? null : description.trim());
        test.setStatus("DRAFT");
        test.setCreatedBy(createdBy);
        test.setCreatedAt(LocalDateTime.now());
        test.setUpdatedAt(LocalDateTime.now());

        int order = 1;
        for (QuestionDraft draft : questions) {
            if (draft.text() == null || draft.text().isBlank()) {
                continue;
            }
            GeneralQuestion question = new GeneralQuestion();
            question.setQuestionText(draft.text().trim());
            question.setQuestionOrder(order++);
            question.setCreatedAt(LocalDateTime.now());
            question.setTest(test);

            int answerOrder = 1;
            for (AnswerDraft answerDraft : draft.answers()) {
                if (answerDraft.text() == null || answerDraft.text().isBlank()) {
                    continue;
                }
                GeneralAnswer answer = new GeneralAnswer();
                answer.setQuestion(question);
                answer.setAnswerLabel(String.valueOf((char) (64 + answerOrder)));
                answer.setAnswerText(answerDraft.text().trim());
                answer.setScore(answerDraft.score());
                answer.setAnswerOrder(answerOrder++);
                answer.setCreatedAt(LocalDateTime.now());
                question.addAnswer(answer);
            }
            test.addQuestion(question);
        }

        if (test.getQuestions().isEmpty()) {
            throw new IllegalArgumentException("Le test doit contenir au moins une question valide");
        }

        return save(test);
    }

    public GeneralTest updateTest(GeneralTest test, String title, String description, String status) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Le titre est obligatoire");
        }
        String trimmedTitle = title.trim();
        if (trimmedTitle.matches("\\d+")) {
            throw new IllegalArgumentException("Le titre ne doit pas être uniquement des nombres");
        }
        if (trimmedTitle.length() < 3 || trimmedTitle.length() > 255) {
            throw new IllegalArgumentException("Le titre doit avoir entre 3 et 255 caractères");
        }
        if (!List.of("DRAFT", "ACTIVE", "INACTIVE", "ARCHIVED").contains(status)) {
            throw new IllegalArgumentException("Statut invalide");
        }

        test.setTitle(trimmedTitle);
        test.setDescription(description == null || description.isBlank() ? null : description.trim());
        test.setStatus(status);
        test.setUpdatedAt(LocalDateTime.now());
        return save(test);
    }

    public record QuestionWithAnswers(GeneralQuestion question, List<GeneralAnswer> answers) {}

    public record QuestionDraft(String text, List<AnswerDraft> answers) {}

    public record AnswerDraft(String text, int score) {}
}
